earth2mip/ensemble_utils.py

```diff
@@ -28,10 +27,146 @@
+
+        # Number of latitudinal modes.
+        self.nlat = nlat
+
+        # Default value of sigma if None is given.
+        if sigma is None:
+            assert alpha > 1.0, f"Alpha must be greater than one, got {alpha}."
+            sigma = tau ** (0.5 * (2 * alpha - 2.0))
+
+        # Inverse SHT
+        self.isht = th.InverseRealSHT(
+            self.nlat, 2 * self.nlat, grid=grid, norm="backward"
+        ).to(dtype=dtype)
+
+        # Square root of the eigenvalues of C.
+        sqrt_eig = (
+            torch.tensor([j * (j + 1) for j in range(self.nlat)])
+            .view(self.nlat, 1)
+            .repeat(1, self.nlat + 1)
+        )
+        sqrt_eig = torch.tril(
+            sigma * (((sqrt_eig / radius**2) + tau**2) ** (-alpha / 2.0))
+        )
+        sqrt_eig[0, 0] = 0.0
+        sqrt_eig = sqrt_eig.unsqueeze(0)
+        self.register_buffer("sqrt_eig", sqrt_eig)
+
+        # Save mean and var of the standard Gaussian.
+        # Need these to re-initialize distribution on a new device.
+        mean = torch.tensor([0.0]).to(dtype=dtype)
+        var = torch.tensor([1.0]).to(dtype=dtype)
+        self.register_buffer("mean", mean)
+        self.register_buffer("var", var)
+
+        # Standard normal noise sampler.
+        self.gaussian_noise = torch.distributions.normal.Normal(self.mean, self.var)
+
+    def forward(self, N, xi=None):
+        """Sample random functions from a spherical GRF.
+
+        Parameters
+        ----------
+        N : int
+            Number of functions to sample.
+        xi : torch.Tensor, default is None
+            Noise is a complex tensor of size (N, nlat, nlat+1).
+            If None, new Gaussian noise is sampled.
+            If xi is provided, N is ignored.
+
+        Output
+        -------
+        u : torch.Tensor
+           N random samples from the GRF returned as a
+           tensor of size (N, nlat, 2*nlat) on a equiangular grid.
+        """
+        # Sample Gaussian noise.
+        if xi is None:
+            xi = self.gaussian_noise.sample(
+                torch.Size((N, self.nlat, self.nlat + 1, 2))
+            ).squeeze()
+            xi = torch.view_as_complex(xi)
+
+        # Karhunen-Loeve expansion.
+        u = self.isht(xi * self.sqrt_eig)
+
+        return u
+
+    # Override cuda and to methods so sampler gets initialized with mean
+    # and variance on the correct device.
+    def cuda(self, *args, **kwargs):
+        super().cuda(*args, **kwargs)
+        self.gaussian_noise = torch.distributions.normal.Normal(self.mean, self.var)
+
+        return self
+
+    def to(self, *args, **kwargs):
+        super().to(*args, **kwargs)
+        self.gaussian_noise = torch.distributions.normal.Normal(self.mean, self.var)
+
+        return self
+
+
 def generate_noise_correlated(shape, *, reddening, device, noise_amplitude):
     return noise_amplitude * brown_noise(shape, reddening).to(device)
 
 
+def generate_noise_grf(shape, grid, alpha, sigma, tau):
```

Comment:
How does this get called from the parsing of the config in say an inference ensemble run